clippy_lints/src/arc_new_in_vec_from_elem.rs

```diff
@@ -0,0 +1,77 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::higher::VecArgs;
+use clippy_utils::last_path_segment;
+use clippy_utils::macros::{root_macro_call_first_node, MacroCall};
+use rustc_hir::{Expr, ExprKind, QPath, TyKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `Arc::new` in `vec![elem; len]`
+    ///
+    /// ### Why is this bad?
+    /// This will create `elem` once and clone it `len` times - doing so with `Arc`
+    /// is a bit misleading, as it will create references to the same pointer, rather
+    /// than different `Arc` instances.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let v = vec![std::sync::Arc::new("some data".to_string()); 100];
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let data = std::sync::Arc::new("some data".to_string());
+    /// let v = vec![data; 100];
+    /// ```
+    #[clippy::version = "1.62.0"]
+    pub ARC_NEW_IN_VEC_FROM_ELEM,
+    suspicious,
+    "calling `Arc::new` in `vec![elem; len]`"
+}
+declare_lint_pass!(ArcNewInVecFromElem => [ARC_NEW_IN_VEC_FROM_ELEM]);
+
+impl LateLintPass<'_> for ArcNewInVecFromElem {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        let Some(macro_call) = root_macro_call_first_node(cx, expr) else { return; };
+
+        if let Some(VecArgs::Repeat(elem, _)) = VecArgs::hir(cx, expr) {
+            if !is_arc_new(cx, elem) {
+                return;
+            }
+
+            yield_lint(cx, &macro_call);
+        }
+    }
+}
+
+fn yield_lint(cx: &LateContext<'_>, macro_call: &MacroCall) {
+    span_lint_and_help(
+        cx,
+        ARC_NEW_IN_VEC_FROM_ELEM,
+        macro_call.span,
+        "calling `Arc::new` in `vec![elem; len]`",
+        None,
+        "consider extracting `Arc` initialization to a variable",
+    );
```

Comment:
```suggestion
    span_lint_and_then(
        cx,
        ARC_NEW_IN_VEC_FROM_ELEM,
        macro_call.span,
        "calling `Arc::new` in `vec![elem; len]`",
        |diag| {
            diag.note("each `Arc` will point to the same allocation");
            diag.help("if this is intentional, consider extracting the `Arc` initialization to a variable");
        },
    );
```

Sorry, one more thing actually :smile:. A note here would be nice to explain what's wrong. Afterwards, could you squash the commits? It should be all good after that.

Reply:
sure thing, looks better